conf/up4.bess

```diff
@@ -0,0 +1,607 @@
+                qspo = __bess_module__("{}QSlowPO".format(name), 'Queue')
+
+                # host_ip_filter: tcpdump -i foo 'dst host 198.19.0.1 or 198.18.0.1' -d
+                # Should always be set to lowest priority
+                HostGate = MAX_GATES - 1
+                ips = ips_by_interface(name)
+                host_ip_filter = {"priority": -HostGate, "filter": "dst host "
+                                + " or ".join(str(x) for x in ips), "gate": HostGate}
+
+                self.bpf.add(filters=[host_ip_filter])
+
+                # Direct control traffic from DPDK to kernel
+                self.bpf:HostGate -> qspo -> spo
+
+                # Direct control traffic from kernel to DPDK
+                spi -> self.fpo
+
+                tc = 'slow{}'.format(wid)
+                try:
+                    bess.add_tc(tc, policy='round_robin', wid=wid)
+                except Exception as e:
+                    if e.errmsg == "Name '{}' already exists".format(tc):
+                        pass
+                    else:
+                        raise e
+                # Limit scheduling slow path RX/TX to 1000 times/second each
+                for mod in spi, qspo:
+                    bess.add_tc(mod.name,
+                            parent=tc,
+                            policy='rate_limit',
+                            resource='count',
+                            limit={'count': 1000})
+                    mod.attach_task(mod.name)
+            except Exception as e:
+                print('Mirror veth interface: {} misconfigured: {}'.format(name, e))
+        else:
+            raise Exception('Invalid mode selected.')
+        # Direct fast path traffic to Frag module
+        merge = __bess_module__("{}Merge".format(name), 'Merge')
+
+        out = self.fpo
+
+        # Attach frag module (if enabled)
+        if ip_frag_with_eth_mtu is not None:
+            frag = __bess_module__("{}IP4Frag".format(name), 'IPFrag', mtu=ip_frag_with_eth_mtu)
+            frag:1 -> out
+            frag:0 -> Sink()
+            out = frag
+
+        # Attach telemeric module (if enabled)
+        if measure:
+            m = Measure()
+            m -> out
+            out = m
+
+        # Attach nat module (if enabled)
+        if self.ext_addrs is not None:
+            # Tokenize the string
+            addrs = self.ext_addrs.split(' or ')
+            # Make a list of ext_addr
+            nat_list = list()
+            for addr in addrs:
+                nat_dict = dict()
+                nat_dict['ext_addr'] = addr
+                nat_list.append(nat_dict)
+
+            # Create the NAT module
+            self.nat = __bess_module__("{}NAT".format(conf[interface]["ifname"]), 'NAT', ext_addrs=nat_list)
+            self.nat:1 -> out
+            out = self.nat
+
+        # Attach Merge module to the 'outlist' of modules
+        merge -> out
+
+
```

Comment:
Should we move this into utils.py or a new file port.py? So that the conf file is more the differentiating stuff?

Reply:
@ajamshed ^ (since comment got overwritten in the files view)

Reply:
I agree. There are 2 ways to get this done:

1. Move the `Port` class to a `ports.py` file. But this will entail creating `bessd` connection and using grpc-based RPC calls via python. This adds up more lines of code.

2. Move the `Port` class to `ports.bess` file. Unfortunately I can't find a way to _import_ bess file (`import` keyword for bess does not work). One method is to insert entire contents of `ports.bess` in `up4.bess` as a macro. This may not look clean.

Reply:
1 feels cleaner and we can leverage as much possible from pybess wrapper.
https://github.com/NetSys/bess/tree/master/pybess